### PR TITLE
Change to the iOS testing option semantics

### PR DESCRIPTION
**Pull request by joerick**
cc @freakboy3742 

I've been thinking about the test-command option, and the implicit addition of `python -m`. 

- I was thinking that the failure mode of any command in here that isn't `pytest` is gonna be pretty confusing for users. e.g. `test-command = "python ./my-tests.py"` -> `python: No module named python`
- it's a bit awkward that if a user's test command really is `python -m mymodule.tests` they have to define it twice, once as-is and again for iOS as `mymodule.tests`

My thought was that we could _require_ test commands on ios start with `python -m`, but with a warning and an auto-fixup where that doesn't occur. That way bad configs always get a warning, and we allow more option reuse between platforms.

I'd want to add a test for this, so just raising as draft for discussion for now.

### CODE REVIEW

**Comment by henryiii** [comment]
Is this required on Pyodide too? I seem to remember `pytest` not working and `python -m pytest` being required. On normal platforms, `pytest` is better than `python -m pytest`, since it does not add the current directory.

**Comment by joerick** [comment]
> Is this required on Pyodide too?

I think it's a bit up in the air, but currently on macOS, `python -m` works better for pyodide also.

**Comment by henryiii** [comment]
When I've added this sort of thing in the past, it wouldn't trigger the expanded tests on the first push (since there's no label there), and then you have to add the label then push something to get the expanded testing.

**Comment by freakboy3742** [comment]
Interesting - I had a similar set of thoughts about this in the process of developing that particular feature. At the time, my thinking went the other direction: preferring the "short" form, and auto-stripping `python -m` from the start of a command line if it was provided. I opted not to do this to avoid Yet Another Weird iOS Behavior; however, I definitely recognise that there's a UX wart here.

The approach you've taken here makes more sense to me, and seems like it would ultimately be more robust. `pytest ...` will work; `python -m pytest ...` will work better; and `./run-test-suite.sh` will fail at runtime in a way that should make *some* sense. So consider this a +1 from me to this idea.

Yes-anding this proposal - there are other predictable error conditions that we could catch in the case where the string doesn't start with `python -m`: anything where the first argument isn't a legal Python module name (i.e., (`all(n.isidentifier() for n in name.split('.'))`) could be immediately flagged as an error, with a similar "must provide a module" helper message.

**Comment by joerick** [comment]
> Yes-anding this proposal - there are other predictable error conditions that we could catch in the case where the string doesn't start with `python -m`: anything where the first argument isn't a legal Python module name (i.e., (`all(n.isidentifier() for n in name.split('.'))`) could be immediately flagged as an error, with a similar "must provide a module" helper message.

Yeah, that's a nice idea. I'd be open to adding it if it's neat.

---

Another thought I had-

What should we do with the test-command if it includes other python flags, e.g. `python -Pm MODULENAME`, which was a pattern we recently discussed in #2358. I can see two options:
1. simply require `python -m` on iOS verbatim - it's not really being executed like that, it's more of a signifier than a real command.
2. Allow users to pass `python -Pm` or `python -P -m`, but don't bother with the other flags. `-P` might become a 'best practice' flag, so it would be nice to reuse config across platforms. `-P` shouldn't be relevant anyway, because test-sources has isolated the test cwd.
3. Try to parse the args more fully, and do _something_ with that...

I think my preference is (1) - it's the most 'honest' version of this, since we'd be discarding any other interpreter flags anyway.

**Comment by freakboy3742** [comment]
> I think my preference is (1) - it's the most 'honest' version of this, since we'd be discarding any other interpreter flags anyway.

I agree. Trying to parse the Python command line is going to be error prone at best. In the context of iOS, other Python command line arguments aren't going to be processed, so doing complex parsing to "pretend" that we are seems like asking for trouble.

**Comment by joerick** [comment]
Whoops! Apologies if you saw some extra commits here, they were meant for a different branch.

File: test/test_ios.py [hunk 1]
```diff
@@ -225,23 +204,74 @@ def test_empty_xbuild_tool_definition(tmp_path, capfd):
...
+        "spam", "0.1.0", platform="ios", python_abi_tags=["cp313-cp313"]
+    )
+    assert set(actual_wheels) == set(expected_wheels)
 
     # The warnings about cross-build notifications were silenced.
     captured = capfd.readouterr()
     assert "Your project configuration does not define any cross-build tools." not in captured.err
+
+
+def test_ios_test_command_without_python_dash_m(tmp_path, capfd):
```
Comment (freakboy3742) [inline test/test_ios.py:217]: If this is actually running the tests, then it should be run in the serial group.
```suggestion
@pytest.mark.serial
def test_ios_test_command_without_python_dash_m(tmp_path, capfd):
```

**Review by freakboy3742 — COMMENTED** [review]
This all makes sense to me - a couple of minor edge cases in testing flagged inline.

File: cibuildwheel/platforms/ios.py [hunk 2]
```diff
@@ -593,14 +593,53 @@ def build(options: Options, tmp_path: Path) -> None:
...
                     log.step("Running test suite...")
+
+                    test_command_parts = shlex.split(build_options.test_command)
+                    if test_command_parts[0:2] != ["python", "-m"]:
+                        first_part = test_command_parts[0]
+                        first_arg_looks_like_a_module_name = all(
+                            n.isidentifier() for n in first_part.split(".")
+                        )
+                        if first_arg_looks_like_a_module_name:
+                            log.warning(
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply cibuildwheel/platforms/ios.py:604]: My counterargument for raising an error on the `python -m`-less form is that `test-command="pytest tests"` is a valid configuration right now for non-mobile platforms, and that configuration will work on iOS. Raising a warning means the user is notified that there's a better option, but the test will *run*.

Yes, it does mean that `notamodule mytest` will be attempted as a test configuration - but that seems an unlikely failure mode in practice, and it *will* fail - just during test execution, with a `ModuleNotFoundError`. The edge case of a "valid shell/exectable that is also a module name and is being used as a valid test command" seems sufficiently niche that it's not worth optimising for that case.

[the remark this answers is not in the document]
Comment (joerick) [reply cibuildwheel/platforms/ios.py:604]: I'd be okay special-casing the allow-and-warn case to only apply to pytest, to allay the concerns around differences between script entrypoint and module. That's gonna be 90% of instances and we know it behaves the same. 

I just like the allow-and-warn pattern from a UX point of view.

[the remark this answers is not in the document]
Comment (mhsmith) [reply cibuildwheel/platforms/ios.py:604]: Yes, I guess pytest at least is worth making a special case, because that will make it easier for many projects to add mobile support with minimal edits to their test configuration.

**Comment by mhsmith** [comment]
Here's [another example](https://github.com/ilanschnell/bitarray/blob/3.3.1/.github/workflows/build_wheels.yml#L30) of a test command pattern which we should probably support, something of the form:
```
python -c "one_liner_which_runs_tests"
```

This could be passed to the testbed and executed using `eval`.

**Comment by freakboy3742** [comment]
> ```
> python -c "one_liner_which_runs_tests"
> ```
> 
> This could be passed to the testbed and executed using `eval`.

Conceptually, sure; but it would need changes in the iOS testbed to support this. The testbed currently assumes a test can be executed as `python -m`.

**Comment by mhsmith** [comment]
In https://github.com/python/cpython/pull/132870 I've generalized the Android testbed script to have `-c` and `-m` arguments, and I'll make cibuildwheel accept any of the following formats on Android:
* `python -c`
* `python -m`
* `pytest` – interpreted as `python -m pytest`

But the bitarray example also raises another question, which is independent of the form of the test command. iOS requires `test-sources` to be non-empty. But this package includes its test suite in the package itself, so empty is actually the correct value.

I can see a few possible solutions:

* Accept an empty value, and copy nothing. Packages which depend on some part of the project directory being available for the tests will fail, possibly in a confusing way.

* Use the same technique as `xbuild-tools` to distinguish between an explicit empty value (which copies nothing), and a default sentinel (which gives an error on mobile, and uses the project directory on desktop).

* Restore your [original fallback approach](https://github.com/pypa/cibuildwheel/pull/2286#discussion_r1976681073) – copying the project directory into the testbed app via an sdist. In the [Android PR](https://github.com/pypa/cibuildwheel/pull/2349), in order to install `build-system` requirements for the build platform,  I've added `build` as a direct dependency of `cibuildwheel` itself. So would now be possible to do this using `[sys.executable, "-m", "build", "--sdist"]`.

I'm leaning towards the last one, because it maximizes the number of packages which will work without modification.

**Comment by freakboy3742** [comment]
> I'm leaning towards the last one, because it maximizes the number of packages which will work without modification.

The downside to the `--sdist` approach is that it could be a little misleading as to whether things can/should be working. For most projects, it will copy a *lot* more than is needed for testing purposes (which could result in slow tests); and will end up with either a *duplicate* copy of the test code (if the project uses a `src` folder), or duplicated entries in the path (if the source tree contains a top level module). In many situations, it might work; but when it doesn't, it's going to fail in weird ways.

Of the options you gave, my preference would be (2) - that you have to specify *something*, but an empty list is an acceptable answer. 

There is, of course, a fourth option that you didn't suggest: "Don't do that" :-) That is, encourage apps to have an actual test suite, rather than embedding test code in app code, by requiring a standalone test suite. That's possibly a *little* opinionated... but I don't think it's an *entirely* unreasonable position to hold.

File: test/test_ios.py [hunk 3]
```diff
@@ -180,21 +170,10 @@ def test_no_xbuild_tool_definition(tmp_path, capfd):
...
-    elif platform_machine == "arm64":
-        expected_wheels = {
-            f"spam-0.1.0-cp313-cp313-ios_{ios_version}_arm64_iphoneos.whl",
-            f"spam-0.1.0-cp313-cp313-ios_{ios_version}_arm64_iphonesimulator.whl",
-        }
-
-    assert set(actual_wheels) == expected_wheels
+    expected_wheels = utils.expected_wheels(
+        "spam", "0.1.0", platform="ios", python_abi_tags=["cp313-cp313"]
+    )
```
[the remark this answers is not in the document]
Comment (joerick) [reply test/test_ios.py:175]: I've added the environment access to expected_wheels. BTW, if this is left unset, does a part of the toolchain set it, or is it determined by the iOS SDK used? I'm just wondering how good of a default "13.0" is.

**Comment by joerick** [comment]
Regarding `python -c`, I don't see a particular need to support that, even the bitarray example can be run just as well via `python -m`. But we can add it if you feel strongly.

Regarding the case of unnecessary `test-sources`. My instinct was to go with the sentinel approach, but we then have this question-

| option value | iOS/Android | other platforms |
|---|---|---|
| _unset_ | Invalid | run from project tree |
| `./some-paths` | run isolated with files | run isolated with files |
| ` ` (empty string) | run isolated without files | ??? |

Probably the answer is to just "run isolated without files". But it means we now lack a way to override to the behaviour "run from project tree". It seems a shame to lose a feature for the sake of a rare edge-case.

I'd be open to 'do nothing'. The user could just workaround this with something silly, like test-sources = "README.md".

I'd also submit a fifth option, a documented special value for test-sources:

5. `test-sources = "<no-files>"` - means 'run isolated without files'. In that case, the table above would look like-

    | option value | iOS/Android | other platforms |
    |---|---|---|
    | _unset_ | Invalid | run from project tree |
    | ` ` (empty string) | Invalid | run from project tree |
    | `./some-paths` | run isolated with files | run isolated with files |
    | `<no-files>` | run isolated without files | run isolated without files |

File: test/test_ios.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply test/test_ios.py]: If it's *completely* unset, then the compiler toolchain falls back to "1.0", which will be compatible with everything... but there's a *lot* missing from iOS SDK 1.0, so it's entirely possible otherwise reasonable C code won't actually compile. This came up working on numpy - if you don't pass through a minimum version when compiling, there are some C++ constructs related to threading that become compilation errors. 

iOS 13.0 is the default version used by the CPython codebase if no version is explicitly specified as part of build configuration (and is also the version used by the BeeWare support package). It's the oldest version that is known to compile Python 3.13 with no build warnings, and targets something like 98% of devices in use today. It's also the version that is [injected into the cibuildwheel environment](https://github.com/pypa/cibuildwheel/blob/main/cibuildwheel/platforms/ios.py#L387) if the user doesn't explicitly specify a minimum version.

CPython on iOS 12.0 *works*, but raises a couple of build warnings about potentially unsupported APIs. iOS 10 is the last version to support 32 bit iPhone devices; it probably also works, but I haven't tested it. Anything older than that is... verging on palaeontology :-)

[the remark this answers is not in the document]
Comment (joerick) [reply test/test_ios.py]: Thanks for the background! Oh right, I had missed that cibuildwheel injects it itself. Great, then we control this default directly.

**Review by henryiii — APPROVED** [review]
